Lab 9: GaussianNB
Jorge Angon, David Beas

Kaggle: https://www.kaggle.com/datasets/nimapourmoradi/raisin-binary-classification

In [1]:
# Raisin classifier

import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder
import numpy as np

# Load the Raisin dataset
# We use a CSV file that contains raisin measurements and class labels

df = pd.read_csv("Raisin_Dataset.csv")   # Load the full dataset

# The dataset columns are:
# Area, MajorAxisLength, MinorAxisLength, Eccentricity,
# ConvexArea, Extent, Perimeter, Class

# X = all numeric measurements
# y = the raisin class labels converted to numbers
X = df[["Area", "MajorAxisLength", "MinorAxisLength",
        "Eccentricity", "ConvexArea", "Extent", "Perimeter"]].values

label_encoder = LabelEncoder()
y = label_encoder.fit_transform(df["Class"])   # Besni=0, Kecimen=1
class_names = label_encoder.classes_

print(f"Total raisins in dataset: {len(X)}")
print(f"Measurements per raisin : {X.shape[1]} (Area, MajorAxisLength, MinorAxisLength, Eccentricity, ConvexArea, Extent, Perimeter)")
print(f"Raisin class names      : {class_names}")
print()

Total raisins in dataset: 900
Measurements per raisin : 7 (Area, MajorAxisLength, MinorAxisLength, Eccentricity, ConvexArea, Extent, Perimeter)
Raisin class names      : ['Besni' 'Kecimen']



In [2]:
# Show the first 5 rows
print("First 5 raisins (measurements):")
print(f"{'Area':>10} {'MajorAxis':>12} {'MinorAxis':>12} {'Eccent.':>10} {'ConvexArea':>12} {'Extent':>10} {'Perimeter':>12} {'Class':>12}")
print("-" * 95)

for i in range(5):
    class_name = class_names[y[i]]
    print(f"{X[i][0]:>10.0f} {X[i][1]:>12.2f} {X[i][2]:>12.2f} {X[i][3]:>10.4f} {X[i][4]:>12.0f} {X[i][5]:>10.4f} {X[i][6]:>12.3f} {class_name:>12}")
print()

First 5 raisins (measurements):
      Area    MajorAxis    MinorAxis    Eccent.   ConvexArea     Extent    Perimeter        Class
-----------------------------------------------------------------------------------------------
     87524       442.25       253.29     0.8197        90546     0.7587     1184.040      Kecimen
     75166       406.69       243.03     0.8018        78789     0.6841     1121.786      Kecimen
     90856       442.27       266.33     0.7984        93717     0.6376     1208.575      Kecimen
     45928       286.54       208.76     0.6850        47336     0.6996      844.162      Kecimen
     79408       352.19       290.83     0.5640        81463     0.7928     1073.251      Kecimen



In [5]:
# Split the data

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,       # 20% for testing = 180 raisins
    random_state=42      # Same split every run (reproducible)
)

print(f"Training raisins: {len(X_train)} (model learns from these)")
print(f"Testing raisins : {len(X_test)}  (model gets graded on these)")
print()

Training raisins: 720 (model learns from these)
Testing raisins : 180  (model gets graded on these)



In [6]:
# Make and train the model
# GaussianNB is used for number data

model = GaussianNB()         # Create an empty model
model.fit(X_train, y_train)  # Train it on 720 raisins

print("Model trained!")
print()

Model trained!



In [7]:
# Test the model on the test data

y_predicted = model.predict(X_test)   # Model guesses each raisin's class

In [8]:
# Check the accuracy

accuracy = accuracy_score(y_test, y_predicted)   # % of correct guesses

print(" How accurate is our model? ")
print(f"Accuracy: {accuracy * 100:.1f}%")
print(f"(Got {int(accuracy * len(y_test))} out of {len(y_test)} raisins correct)")
print()

 How accurate is our model? 
Accuracy: 82.8%
(Got 149 out of 180 raisins correct)



In [9]:
# Show the confusion matrix
# This shows right and wrong guesses

cm = confusion_matrix(y_test, y_predicted)

print("Confusion Matrix")
print("Rows = actual class, Columns = predicted class")
print(f"{'':15} {'Besni':>10} {'Kecimen':>10}")
print("-" * 37)

for i, row in enumerate(cm):
    print(f"{class_names[i]:15} {row[0]:>10} {row[1]:>10}")
print()

Confusion Matrix
Rows = actual class, Columns = predicted class
                     Besni    Kecimen
-------------------------------------
Besni                   63         23
Kecimen                  8         86



In [10]:
# Predict one new raisin

print("Predict a brand new raisin")
print()

# A new raisin with measurements
new_raisin = [[87524, 442.25, 253.29, 0.8200, 90546, 0.7587, 1184.040]]
# [Area, MajorAxisLength, MinorAxisLength, Eccentricity, ConvexArea, Extent, Perimeter]

# Get the prediction
prediction = model.predict(new_raisin)[0]                  # 0 or 1
class_name = class_names[prediction]                       # Convert to class name
probabilities = model.predict_proba(new_raisin)[0]         # Probability for each class

print(f"New raisin measurements:")
print(f"  Area            : {new_raisin[0][0]}")
print(f"  MajorAxisLength : {new_raisin[0][1]}")
print(f"  MinorAxisLength : {new_raisin[0][2]}")
print(f"  Eccentricity    : {new_raisin[0][3]}")
print(f"  ConvexArea      : {new_raisin[0][4]}")
print(f"  Extent          : {new_raisin[0][5]}")
print(f"  Perimeter       : {new_raisin[0][6]}")
print()
print(f"Prediction: {class_name.upper()}")
print()
print(f"Probabilities:")
for i, name in enumerate(class_names):
    bar = "#" * int(probabilities[i] * 30)    # Simple text bar chart
    print(f"  {name:12} | {bar:<30} | {probabilities[i]*100:.1f}%")

Predict a brand new raisin

New raisin measurements:
  Area            : 87524
  MajorAxisLength : 442.25
  MinorAxisLength : 253.29
  Eccentricity    : 0.82
  ConvexArea      : 90546
  Extent          : 0.7587
  Perimeter       : 1184.04

Prediction: KECIMEN

Probabilities:
  Besni        | ##############                 | 49.0%
  Kecimen      | ###############                | 51.0%


For this dataset, we used Gaussian Naive Bayes to classify raisins into two groups, Besni and Kecimen, using numeric measurements like area, axis lengths, eccentricity, convex area, extent, and perimeter. We loaded the dataset, split it into training and testing data, trained the model, and checked its performance with accuracy and a confusion matrix.

The model reached 82.8% accuracy which shows that it correctly classified 149 out of 180 test raisins. The confusion matrix showed that it predicted most raisins correctly, but it still made some mistakes between the two classes, which is normal when the measurements are similar. We also tested one brand new raisin entry, and the model predicted Kecimen with 51.0% probability, showing it can make predictions on new data, not just the rows it trained on.